Improved efficiency of Members request caching for the Posts Content API

allouis opened the pull request:
At the moment we're using the entire `option` object as the cache key for requests to the posts content API. For unauthenticated requests this is fairly okay, but for requests with members authentication we are storing the same content under many keys!

The changes here are to enable custom key data generation at the endpoint layer, as well as implementing custom key data for the posts api.

We strip all options which don't have an effect on the content, e.g. `debug`. And then only include member data that has an effect on the content. This reduces the number of unique keys and so allows us to cache more data.

allouis committed:
Added support for custom cache key generation

refs https://github.com/TryGhost/Arch/issues/83

This allows endpoints to implement their own key generation, with access to the
frame object they can be smart about key generation and use only options and
context values that are appropriate.

allouis committed:
Customised key generation for Posts Content API

refs https://github.com/TryGhost/Arch/issues/83

The options included are the only ones which have an effect on the response
data, as well as that we are using the properties of members which are used by
content-gating module. For the read operation we need to include the ID too.

naz commented on ghost/api-framework/lib/pipeline.js:
```diff
@@ -230,6 +230,9 @@ const pipeline = (apiController, apiUtils, apiType) => {
             frame.method = method;
 
             let cacheKey = JSON.stringify(frame.options);
```
We can save some time doing this serialization only when needed. `JSON.stringify` is not a crazy expensive operation but who knows what's in the frame options and would add up unnecessarily 

What I mean is:
```
let cacheKey;

if (apiImpl.generateCacheKey)
   cacheKey = await apiImpl.generateCacheKey(frame);
} else {
  // default key
  cacheKey = JSON.stringify(frame.options);
}

allouis replied:
Yup! And can also use the stable-json-stringify here too

naz replied:
Agreed!

naz commented on ghost/core/core/server/api/endpoints/posts-public.js:
```diff
@@ -31,6 +31,34 @@ module.exports = {
+                formats: frame.options?.formats?.split(',').sort(),
+                limit: frame.options?.limit,
+                order: frame.options?.order,
+                page: frame.options?.page || 1,
```
Interesting one. Should we consider setting a default for the `options.page` on some other layer to avoid the `||` condition here? 🤔

allouis replied:
Potentially yeah, but it's tricky because this gets run before most other layers.

allouis replied:
It doesn't really work unless we shift lots of things round, and start running serialisation before cache key generation. Leaving as-is for now

naz commented on ghost/core/core/server/api/endpoints/posts-public.js:
```diff
@@ -31,6 +31,34 @@ module.exports = {
+                    tiers: frame.options?.context?.member.products?.map((product) => {
+                        return product.slug;
+                    }).sort(),
+                    labels: frame.options?.context.member.labels?.map((label) => {
```
Do labels affect content gating?

allouis replied:
Nope! As discussed in slack - have removed :)

allouis committed:
Restricted content gating filter to status & tiers

refs https://github.com/TryGhost/Arch/issues/83

As we've only used the status and tiers in the cache key generation (as those
are currently the only ones that are present in DB's) we want to make sure that
content gating doesn't use any other properties, which means the cache behaviour
will match the content gating behaviour
